### PR TITLE
Python: add sinter command

### PR DESCRIPTION
**Pull request by yipin-chen**
https://redis.io/docs/latest/commands/sinter/

### CODE REVIEW

File: python/python/glide/async_commands/core.py [hunk 1]
```diff
@@ -1527,6 +1527,26 @@ async def sismember(
...
+
+        Args:
+            keys (List[str]): The keys of the sets.
+
+        Returns:
+            Set[str]: A set of members which are present in all given sets.
+                Missing or empty input sets cause an empty set response.
+
+        Examples:
+            >>> await client.sadd("my_set1", ["member1", "member2"])
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:1543]: nit: you don't need to include other commands, just describe what's in the Set via a comment.

File: python/python/glide/async_commands/core.py [hunk 2]
```diff
@@ -1527,6 +1527,26 @@ async def sismember(
...
+            >>> await client.sadd("my_set2", ["member2", "member3"])
+            >>> await client.sinter(["my_set1", "my_set2"])
+                 "member2" # matched member
+        """
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:1547]: Java includes an example using a non-existing set:
```
     * Set<String> values = client.sinter(new String[] {"set1", "nonExistingSet"}).get();
     * assert values.size() == 0;
```

File: python/python/glide/async_commands/core.py [hunk 3]
```diff
@@ -1527,6 +1527,26 @@ async def sismember(
...
+
+        Returns:
+            Set[str]: A set of members which are present in all given sets.
+                Missing or empty input sets cause an empty set response.
+
+        Examples:
+            >>> await client.sadd("my_set1", ["member1", "member2"])
+            >>> await client.sadd("my_set2", ["member2", "member3"])
+            >>> await client.sinter(["my_set1", "my_set2"])
+                 "member2" # matched member
```
[the remark this answers is not in the document]
Comment (acarbonetto) [reply python/python/glide/async_commands/core.py:1546]:
```
{"member2"}
```

File: python/python/tests/test_async_client.py [hunk 4]
```diff
@@ -1107,6 +1107,19 @@ async def test_spop(self, redis_client: TRedisClient):
...
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_sinter(self, redis_client: TRedisClient):
+        key1 = f"{{testKey}}:{get_random_string(10)}"
+        key2 = f"{{testKey}}:{get_random_string(10)}"
+        member1_list = ["a", "b", "c"]
+        member2_list = ["c", "d", "e"]
+
+        assert await redis_client.sadd(key1, member1_list) == 3
+        assert await redis_client.sadd(key2, member2_list) == 3
+        # assert await redis_client.custom_command(["SINTER", key1, key2]) == {"c"}
```
Comment (acarbonetto) [inline python/python/tests/test_async_client.py:1120]: remove?

File: python/python/tests/test_async_client.py [hunk 5]
```diff
@@ -1107,6 +1107,19 @@ async def test_spop(self, redis_client: TRedisClient):
...
+        assert await redis_client.sinter([key1, key2]) == {"c"}
+
```
Comment (acarbonetto) [inline python/python/tests/test_async_client.py:1122]: Please add a test with a non-existing key
Please add a test with a non-Set key

Comment (acarbonetto) [reply python/python/tests/test_async_client.py:1122]: https://github.com/aws/glide-for-redis/blob/main/java/integTest/src/test/java/glide/SharedCommandTests.java#L1255-L1268

File: python/python/glide/async_commands/core.py (hunk 2 above)
Comment (Yury-Fridlyand) [reply python/python/glide/async_commands/core.py:1547]: a bit overkill though

File: python/python/glide/async_commands/core.py [hunk 6]
```diff
@@ -1527,6 +1527,26 @@ async def sismember(
             await self._execute_command(RequestType.SIsMember, [key, member]),
         )
 
+    async def sinter(self, keys: List[str]) -> Set[str]:
+        """
+        Gets the intersection of all the given sets.
+        See https://redis.io/docs/latest/commands/sinter/ for more details.
+
+        Args:
```
Comment (Yury-Fridlyand) [inline python/python/glide/async_commands/core.py:1535]: Add same-slot notice (see #282 for example)
> When in Cluster mode, all keys must map to the same hash slot.

File: python/python/tests/test_async_client.py (hunk 5 above)
Comment (Yury-Fridlyand) [inline python/python/tests/test_async_client.py:1122]: Add a test for cross-slot case.
Note: it is being reworked in https://github.com/aws/glide-for-redis/pull/1391

**Comment by Yury-Fridlyand** [comment]
Update change log when you'll be posting that on upstream.

File: python/python/glide/async_commands/core.py [hunk 7]
```diff
@@ -1527,6 +1527,26 @@ async def sismember(
...
+            keys (List[str]): The keys of the sets.
+
+        Returns:
+            Set[str]: A set of members which are present in all given sets.
+                Missing or empty input sets cause an empty set response.
```
[the remark this answers is not in the document]
Comment (yipin-chen) [reply python/python/glide/async_commands/core.py:1540]: I have updated Python-side according to your suggestion.
Do I apply the same change to Java-side with this same PR?

File: python/python/glide/async_commands/core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yipin-chen) [reply python/python/glide/async_commands/core.py:1546]: Nice catch. I have updated.

File: python/python/glide/async_commands/transaction.py [hunk 8]
```diff
@@ -1010,6 +1010,21 @@ def sismember(
...
+        Gets the intersection of all the given sets.
+
+        See https://redis.io/docs/latest/commands/sinter/ for more details.
+
+        Args:
+            keys (List[str]): The keys of the sets.
+
+        Command response:
+            Set[str]: A set of members which are present in all given sets.
+                Missing or empty input sets cause an empty set response.
```
[the remark this answers is not in the document]
Comment (yipin-chen) [reply python/python/glide/async_commands/transaction.py:1024]: I have updated according to above suggestion.

File: python/python/tests/test_async_client.py (hunk 4 above)
Comment (yipin-chen) [reply python/python/tests/test_async_client.py:1120]: Will do.

File: python/python/glide/async_commands/core.py (hunk 6 above)
Comment (yipin-chen) [reply python/python/glide/async_commands/core.py:1535]: Will do.

File: python/python/glide/async_commands/core.py (hunk 1 above)
Comment (yipin-chen) [reply python/python/glide/async_commands/core.py:1543]: I just follow other examples.

File: python/python/tests/test_transaction.py [hunk 9]
```diff
@@ -203,6 +205,12 @@ async def transaction_test(
     args.append(2)
     transaction.spop_count(key7, 4)
     args.append({"foo", "bar"})
+    transaction.sadd(key16, ["a", "b", "c"])
+    args.append(3)
+    transaction.sadd(key17, ["c", "d", "e"])
+    args.append(3)
+    transaction.sinter([key16, key17])
```
[the remark this answers is not in the document]
Comment (yipin-chen) [reply python/python/tests/test_transaction.py:212]: Change to use key7 twice.

**Comment by Yury-Fridlyand** [comment]
Please refresh both branches to avoid non related changes

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,123 +1,130 @@
 #### Changes
-* Python: Added JSON.DEL JSON.FORGET commands  ([#1146](https://github.com/aws/glide-for-redis/pull/1146))
```
Comment (aaron-congo) [inline CHANGELOG.md:2]: As discussed, looks like there were a few changes accidentally included here, there should only be one extra line adding the notice about SINTER in this file

File: python/python/glide/async_commands/core.py [hunk 11]
```diff
@@ -1601,6 +1601,29 @@ async def sunionstore(
             await self._execute_command(RequestType.SUnionStore, [destination] + keys),
         )
 
+    async def sinter(self, keys: List[str]) -> Set[str]:
+        """
+        Gets the intersection of all the given sets.
+        When in cluster mode, all keys must map to the same hash slot.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/core.py:1607]: Can you add a blank line between this line and the one above, also we recently changed the format of this notice, can you change to align with the new format (should have "Note:" above)? You can see an example [here](https://github.com/aws/glide-for-redis/blob/main/python/python/glide/async_commands/core.py#L3247)

File: python/python/glide/async_commands/core.py [hunk 12]
```diff
@@ -1601,6 +1601,29 @@ async def sunionstore(
...
+        Returns:
+            Set[str]: A set of members which are present in all given sets.
+                If one or more sets do no exist, an empty set will be returned.
+
+        Examples:
+            >>> await client.sadd("my_set1", ["member1", "member2"])
+            >>> await client.sadd("my_set2", ["member2", "member3"])
+            >>> await client.sinter(["my_set1", "my_set2"])
+                 {"member2"} # matched member
+            >>> await client.sinter([my_set1", "non_exiting_set"])
```
Comment (aaron-congo) [inline python/python/glide/async_commands/core.py:1622]: small typo
```suggestion
            >>> await client.sinter([my_set1", "non_existing_set"])
```

File: python/python/glide/async_commands/transaction.py [hunk 13]
```diff
@@ -1052,6 +1052,21 @@ def sunionstore(
         """
         return self.append_command(RequestType.SUnionStore, [destination] + keys)
 
+    def sinter(self: TTransaction, keys: List[str]) -> TTransaction:
+        """
+        Gets the intersection of all the given sets.
+
+        See https://redis.io/docs/latest/commands/sinter/ for more details.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/transaction.py:1059]:
```suggestion
        See https://valkey.io/docs/latest/commands/sinter for more details.
```

File: python/python/tests/test_async_client.py [hunk 14]
```diff
@@ -1190,6 +1190,29 @@ async def test_sunionstore(self, redis_client: TRedisClient):
...
+        member2_list = ["c", "d", "e"]
+
+        # positive test case
+        assert await redis_client.sadd(key1, member1_list) == 3
+        assert await redis_client.sadd(key2, member2_list) == 3
+        assert await redis_client.sinter([key1, key2]) == {"c"}
+
+        # non-existing key returns empty set
+        assert await redis_client.sinter([key1, non_existing_key]) == set()
+
```
Comment (aaron-congo) [inline python/python/tests/test_async_client.py:1209]: Sho has asked me before to add a test where you pass an empty key list to the command. The test should assert that the command fails since the key list should not be empty. Can you add that test here? You can use this as an example: https://github.com/aws/glide-for-redis/blob/main/python/python/tests/test_async_client.py#L2099

File: python/python/glide/async_commands/core.py [hunk 15]
```diff
@@ -1601,6 +1601,32 @@ async def sunionstore(
...
+        """
+        Gets the intersection of all the given sets.
+
+        See https://valkey.io/docs/latest/commands/sinter/ for more details.
+
+        Args:
+            keys (List[str]): The keys of the sets.
+
+        Note:
+            When in cluster mode, all 'keys' must map to the same hash slot.
```
Comment (acarbonetto) [inline python/python/glide/async_commands/core.py:1614]:
```suggestion
            When in cluster mode, all `keys` must map to the same hash slot.
```

File: python/python/tests/test_transaction.py [hunk 16]
```diff
@@ -209,6 +209,8 @@ async def transaction_test(
     args.append({"foo", "bar"})
     transaction.smove(key7, key7, "non_existing_member")
     args.append(False)
+    transaction.sinter([key7, key7])
+    args.append(set())
```
Comment (acarbonetto) [inline python/python/tests/test_transaction.py:213]: is key7 empty at this point?  if you move this up to be after sunionstore (after line 206) you should be able to do sinter with a return value.

**Review by acarbonetto — COMMENTED** [review]
minor comments - otherwise, looks good!

File: python/python/tests/test_transaction.py (hunk 16 above)
Comment (yipin-chen) [reply python/python/tests/test_transaction.py:213]: Will move right after line 206.

File: python/python/glide/async_commands/core.py [hunk 17]
```diff
@@ -1601,6 +1601,32 @@ async def sunionstore(
             await self._execute_command(RequestType.SUnionStore, [destination] + keys),
         )
 
+    async def sinter(self, keys: List[str]) -> Set[str]:
+        """
+        Gets the intersection of all the given sets.
+
+        See https://valkey.io/docs/latest/commands/sinter/ for more details.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/core.py:1608]: nit:
```suggestion
        See https://valkey.io/docs/latest/commands/sinter for more details.
```

File: python/python/glide/async_commands/transaction.py [hunk 18]
```diff
@@ -1052,6 +1052,21 @@ def sunionstore(
         """
         return self.append_command(RequestType.SUnionStore, [destination] + keys)
 
+    def sinter(self: TTransaction, keys: List[str]) -> TTransaction:
+        """
+        Gets the intersection of all the given sets.
+
+        See https://valkey.io/docs/latest/commands/sinter/ for more details.
```
Comment (aaron-congo) [inline python/python/glide/async_commands/transaction.py:1059]: nit:
```suggestion
        See https://valkey.io/docs/latest/commands/sinter for more details.
```

**Review by aaron-congo — APPROVED** [review]
